src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableInterlocked.cs

```diff
@@ -12,6 +12,92 @@ namespace System.Collections.Immutable
     /// </summary>
     public static class ImmutableInterlocked
     {
+        /// <summary>
+        /// Mutates a value in-place with optimistic locking transaction semantics
+        /// via a specified transformation function.
+        /// The transformation is retried as many times as necessary to win the optimistic locking race.
+        /// </summary>
+        /// <typeparam name="T">The type of data.</typeparam>
+        /// <param name="location">
+        /// The variable or field to be changed, which may be accessed by multiple threads.
+        /// </param>
+        /// <param name="transformer">
+        /// A function that mutates the value. This function should be side-effect free, 
+        /// as it may run multiple times when races occur with other threads.</param>
+        /// <returns>
+        /// <c>true</c> if the location's value is changed by applying the result of the 
+        /// <paramref name="transformer"/> function;
+        /// <c>false</c> if the location's value remained the same because the last 
+        /// invocation of <paramref name="transformer"/> returned the existing value.
+        /// </returns>
+        public static bool ApplyChange<T>(ref T location, Func<T, T> transformer) where T : class
```

Comment:
I'd vote for either `ApplyChange` (which would be exact) or `Update` (which would be less specific, but easier on the eyes). I do not like `AssignValue` because the method is really not just assigning a value.